cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -90,181 +129,223 @@ func newCmdApply(apf *applyPlanFlags) *cobra.Command {
+	if !ok {
+		return nil, cmdutil.TypeMismatchErr("allowExperimentalUpgrades", "bool")
 	}
 
-	// If the current session is interactive, ask the user whether they really want to upgrade.
-	dryRun, ok := cmdutil.ValueFromFlagsOrConfig(flagSet, options.DryRun, upgradeCfg.Apply.DryRun, &flags.dryRun).(*bool)
-	if ok {
-		flags.dryRun = *dryRun
-	} else {
-		return cmdutil.TypeMismatchErr("dryRun", "bool")
+	allowRCUpgrades, ok := cmdutil.ValueFromFlagsOrConfig(cmd.Flags(), "allow-release-candidate-upgrades", upgradeCfg.Apply.AllowRCUpgrades, &applyFlags.allowRCUpgrades).(*bool)
+	if !ok {
+		return nil, cmdutil.TypeMismatchErr("allowRCUpgrades", "bool")
 	}
 
-	if flags.sessionIsInteractive() {
-		if err := cmdutil.InteractivelyConfirmAction("upgrade", "Are you sure you want to proceed?", os.Stdin); err != nil {
-			return err
-		}
+	printConfig, ok := cmdutil.ValueFromFlagsOrConfig(cmd.Flags(), "print-config", upgradeCfg.Apply.PrintConfig, &applyFlags.printConfig).(*bool)
+	if !ok {
+		return nil, cmdutil.TypeMismatchErr("printConfig", "bool")
 	}
 
-	if !flags.dryRun {
-		fmt.Println("[upgrade/prepull] Pulling images required for setting up a Kubernetes cluster")
-		fmt.Println("[upgrade/prepull] This might take a minute or two, depending on the speed of your internet connection")
-		fmt.Println("[upgrade/prepull] You can also perform this action beforehand using 'kubeadm config images pull'")
-		if err := preflight.RunPullImagesCheck(utilsexec.New(), initCfg, sets.New(upgradeCfg.Apply.IgnorePreflightErrors...)); err != nil {
-			return err
-		}
-	} else {
-		fmt.Println("[upgrade/prepull] Would pull the required images (like 'kubeadm config images pull')")
+	client, err := getClient(applyFlags.kubeConfigPath, *dryRun)
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't create a Kubernetes client from file %q", applyFlags.kubeConfigPath)
 	}
 
-	waiter := getWaiter(flags.dryRun, client, upgradeCfg.Timeouts.UpgradeManifests.Duration)
+	printer := &output.TextPrinter{}
 
-	// If the config is set by flag, just overwrite it!
-	etcdUpgrade, ok := cmdutil.ValueFromFlagsOrConfig(flagSet, options.EtcdUpgrade, upgradeCfg.Apply.EtcdUpgrade, &flags.etcdUpgrade).(*bool)
-	if ok {
-		upgradeCfg.Apply.EtcdUpgrade = etcdUpgrade
-	} else {
-		return cmdutil.TypeMismatchErr("etcdUpgrade", "bool")
+	// Fetches the cluster configuration
+	klog.V(1).Infoln("[upgrade/apply] retrieving configuration from cluster")
+	initCfg, err := configutil.FetchInitConfigurationFromCluster(client, nil, "upgrade", false, false)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			_, _ = printer.Printf("[upgrade/config] In order to upgrade, a ConfigMap called %q in the %q namespace must exist.\n", constants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
+			_, _ = printer.Printf("[upgrade/config] Use 'kubeadm init phase upload-config --config your-config.yaml' to re-upload it.\n")
```

Comment:
```suggestion
			_, _ = printer.Printf("[upgrade] In order to upgrade, a ConfigMap called %q in the %q namespace must exist.\n", constants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
			_, _ = printer.Printf("[upgrade] Use 'kubeadm init phase upload-config --config your-config.yaml' to re-upload it.\n")
```

trying to avoid confusing now that we have phases, but "config " is not a phase.